internal/transport/http2_client.go

```diff
@@ -1006,24 +1008,35 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
+
+	var st *status.Status
 	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
-	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY. It
+	// also waits for loopyWriter to be closed with a timer to avoid the
+	// long blocking in case the connection is blackholed, i.e. TCP is
+	// just stuck.
 	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("client transport shutdown"), closeConn: err})
-	<-t.writerDone
+	timer := time.NewTimer(goAwayLoopyWriterTimeout)
+	defer timer.Stop()
+	select {
+	case <-t.writerDone:
+		// Append info about previous goaway's if there were any, since this
+		// may be important for understanding the root cause for this
+		// connection to be closed.
+		_, goAwayDebugMessage := t.GetGoAwayReason()
+		if len(goAwayDebugMessage) > 0 {
+			st = status.Newf(codes.Unavailable, "closing transport due to: %v, received prior goaway: %v", err, goAwayDebugMessage)
+			err = st.Err()
+		} else {
+			st = status.New(codes.Unavailable, err.Error())
+		}
+	case <-timer.C:
+		st = status.New(codes.Unavailable, err.Error())
```

Comment:
Wait.. so if the timer fires we're losing the goaway debugging info.  That's not ideal.

Why did you couple this error logic with the timer/writerDone stuff?  I don't think that's needed.  Can you leave the error handling stuff alone and just replace `<-t.writerDone` with:
```go
select {
	case <-t.writerDone: // success
	case <-timer.C:
		t.logger.Info(...)
}
```

Reply:
This looks much better now, thanks..